### PR TITLE
Add support for datadog query function wrappers - WIP

### LINKED ISSUE #2761: Datadog query with functions never returns healthy
### Report

Edit: see comments below, also amended the Issue title to better match the issue at hand

I have a Datadog ScaledObject resource setup, which looks like the following (Secret called ```keda-myapp-datadog-secret``` already created):
```
---
apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: myapp
  namespace: myapp
spec:
  secretTargetRef:
  - parameter: apiKey
    name: keda-myapp-datadog-secret
    key: apiKey
  - parameter: appKey
    name: keda-myapp-datadog-secret
    key: appKey
---
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: myapp
  namespace: myapp
spec:
  scaleTargetRef:
    name: myapp
  minReplicaCount:  1
  maxReplicaCount:  3
  advanced:
    horizontalPodAutoscalerConfig:
      behavior:
        scaleDown:
          stabilizationWindowSeconds: 600
          policies:
          - type: Pods
            value: 1
            periodSeconds: 60
  triggers:
  - type: datadog
    metadata:
      query: "per_second(sum:http.requests{service:myapp,dc:us-west-2}.rollup(avg, 60))"
      queryValue: "100"
      type: "Average"
      age: "300"
    authenticationRef:
      name: myapp
  - type: datadog
    metadata:
      query: "sum:http.backlog{service:myapp,dc:us-west-2}.rollup(avg, 60)"
      queryValue: "50"
      type: "Average"
      age: "300"
    authenticationRef:
      name: myapp
```
Note specifically, the two datadog triggers in the single ScaledObject.

With this setup, the scaled object reports the following Health state (from describe):
```
External Metric Names:                                                                                                                                                                                             
    s0-datadog-per_second(sum-http-requests
    s1-datadog-sum-http-backlog
Health:
    s0-datadog-per_second(sum-http-requests:
        Number Of Failures:  63
        Status:              Failing
    s1-datadog-sum-http-backlog:
        Number Of Failures:  0
        Status:              Happy
```
Note that the first Trigger is failing, whilst the second is Happy.

And the keda-operator-metrics-apiserver logs show:
```
E0315 11:01:52.364068       1 logr.go:270] keda_metrics_adapter/datadog_scaler "msg"="error getting metrics from Datadog" "error"="no Datadog metrics returned"
E0315 11:01:52.364168       1 scale_resolvers.go:164] keda_metrics_adapter/scalehandler "msg"="Error getting triggerAuth" "error"="context canceled" "name"="myapp" "namespace"="myapp" "type"="ScaledObject" "triggerAuthRef.Name"="myapp"
E0315 11:01:52.376993       1 provider.go:124] keda_metrics_adapter/provider "msg"="error getting metric for scaler" "error"="error parsing Datadog metadata: no api key given"  "scaledObject.Name"="myapp" "scaledObject.Namespace"="myapp" "scaler"={}
E0315 11:01:52.377048       1 status.go:71] apiserver received an error that is not an metav1.Status: &errors.errorString{s:"no matching metrics found for s0-datadog-per_second(sum-http-requests"}: no matching metrics found for s0-datadog-per_second(sum-http-requests
I0315 11:01:53.322901       1 trace.go:205] Trace[1690803741]: "List" url:/apis/external.metrics.k8s.io/v1beta1/namespaces/myapp/s1-datadog-sum-http-backlog,user-agent:kube-controller-manager/v1.21.5 (linux/amd64) kubernetes/5236faf/system:serviceaccount:kube-system:horizontal-pod-autoscaler,audit-id:694c8fc1-3d50-455d-9bb5-e22514569a2a,client:172.16.41.151,accept:application/vnd.kubernetes.protobuf, */*,protocol:HTTP/2.0 (15-Mar-2022 11:01:52.380) (total time: 942ms):
Trace[1690803741]: ---"Listing from storage done" 941ms (11:01:53.322)
Trace[1690803741]: [942.018779ms] [942.018779ms] END
```
Note here, that it's reporting ```no api key given``` even though it's using the exact same ```authenticationRef``` value to the other trigger.

### Expected Behavior

When:
* Multiple Datadog triggers defined on a single ScaledObject
* All triggers using the same TriggerAuth object
* Trigger Auth loaded from Kubernetes Secret

All Triggers should load metrics successfully.

### Actual Behavior

Only the last trigger correctly loads the configured TriggerAuth. Other triggers report ```no api key given```

### Steps to Reproduce the Problem

See report for example YAML

### Logs from KEDA operator

See above.

### KEDA Version

2.6.1

### Kubernetes Version

1.21

### Platform

Amazon Web Services

### Scaler Details

Datadog

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by dalgibbard**
Allow the Datadog Scaler to parse/validate queries which contain a wrapping function; eg:
```
per_second(sum:system.cpu.user{*}.rollup(avg, 30))
```
Previously, the code would see these as missing an aggregator; and prefix them with ```avg:``` resulting in an invalid query of:
```
avg:per_second(sum:system.cpu.user{*}.rollup(avg, 30))
```

### Checklist

- [X] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [X] Tests have been added
- [ ] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [ ] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [X] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)

Fixes #2761

### CODE REVIEW

File: pkg/scalers/datadog_scaler.go [hunk 1]
```diff
@@ -46,12 +46,14 @@ type datadogMetadata struct {
 
 var datadogLog = logf.Log.WithName("datadog_scaler")
 
-var aggregator, filter, rollup *regexp.Regexp
+var aggregator, filter, rollup, ddfuncpresent, ddfunc *regexp.Regexp
 
 func init() {
-	aggregator = regexp.MustCompile(`^(avg|sum|min|max):.*`)
+	aggregator = regexp.MustCompile(`^([\w]+\()?(avg|sum|min|max):.*`)
```
Comment (zroubalik) [inline pkg/scalers/datadog_scaler.go:52]: Pardon my ignorance, just looking at all these regexes, isn't there any validator in any Datadog library?

Comment (dalgibbard) [reply pkg/scalers/datadog_scaler.go:52]: An excellent question; will check!

Comment (dalgibbard) [reply pkg/scalers/datadog_scaler.go:52]: Had a quick look around, plenty for submitting metrics etc, but not a lot in the way of validating if a metrics query is valid.

The Datadog UI posts to https://app.datadoghq.com/api/v2/query/timeseries when changing queries interactively, and returns 400 when it's invalid - but seeing as this requires auth anyway, if going down that route, you may as well skip validation in the Scaler, and just handle errors when they come back from the upstream.

**Comment by dalgibbard** [comment]
Closing this, as it makes sense to not validate in the code, and instead use the upstream API.

**Pull request closed without merging**